lib/config/config-file.js

```diff
@@ -0,0 +1,366 @@
+    if (!Array.isArray(config.extends)) {
+        configExtends = [config.extends];
+    }
+
+    // Make the last element in an array take the highest precedence
+    config = configExtends.reduceRight(function(previousValue, parentPath) {
+
+        if (parentPath === "eslint:recommended") {
+            // Add an explicit substitution for eslint:recommended to conf/eslint.json
+            // this lets us use the eslint.json file as the recommended rules
+            parentPath = path.resolve(__dirname, "../../conf/eslint.json");
+        } else if (isFilePath(parentPath)) {
+            // If the `extends` path is relative, use the directory of the current configuration
+            // file as the reference point. Otherwise, use as-is.
+            parentPath = (!isAbsolutePath(parentPath) ?
+                path.join(path.dirname(filePath), parentPath) :
+                parentPath
+            );
+        }
+
+        try {
+            debug("Loading " + parentPath);
+            return ConfigOps.merge(load(parentPath), previousValue);
+        } catch (e) {
+            // If the file referenced by `extends` failed to load, add the path to the
+            // configuration file that referenced it to the error message so the user is
+            // able to see where it was referenced from, then re-throw
+            e.message += "\nReferenced from: " + filePath;
+            throw e;
+        }
+
+    }, config);
+
+    return config;
+}
+
+/**
+ * Resolves a configuration file path into the fully-formed path, whether filename
+ * or package name.
+ * @param {string} filePath The filepath to resolve.
+ * @returns {string} A path that can be used directly to load the configuration.
+ * @private
+ */
+function resolve(filePath) {
+
+    if (isFilePath(filePath)) {
+        return path.resolve(filePath);
+    } else {
+
+        // it's a package
+
+        if (filePath.charAt(0) === "@") {
+            // it's a scoped package
+
+            // package name is "eslint-config", or just a username
+            var scopedPackageShortcutRegex = /^(@[^\/]+)(?:\/(?:eslint-config)?)?$/;
+            if (scopedPackageShortcutRegex.test(filePath)) {
+                filePath = filePath.replace(scopedPackageShortcutRegex, "$1/eslint-config");
+            } else if (filePath.split("/")[1].indexOf("eslint-config-") !== 0) {
+                // for scoped packages, insert the eslint-config after the first /
+                filePath = filePath.replace(/^@([^\/]+)\/(.*)$/, "@$1/eslint-config-$2");
+            }
+        } else if (filePath.indexOf("eslint-config-") !== 0) {
+            filePath = "eslint-config-" + filePath;
+        }
+
+        return filePath;
+    }
+
+}
+
+/**
+ * Loads a configuration file from the given file path.
+ * @param {string} filePath The filename or package name to load the configuration
+ *      information from.
+ * @returns {Object} The configuration information.
+ * @private
+ */
+function load(filePath) {
```

Comment:
Can this function be moved up, so you wouldn't have to disable `no-use-before-define` for the entire file?